src/menhir_parser.mly

```diff
@@ -4,16 +4,107 @@ open Lang
   %}
 
 %token EOF
+%token COLUM
+%token SEMICOL
+%token EQUAL
+%token DASH
+%token<string> STR
+%token<string> URL
+%token<int> INT
 
 %start <Lang.identifier> identifier
 
 %%
   let identifier_core :=
-      | { (1, Snapshot, [||]) }
+      | swh = STR; COLUM; scheme_v = INT; COLUM; obj_t = STR; COLUM; hash = STR; {
 
-let qualifiers :=
-    | { [] }
+    if swh <> "swh" then
+      failwith "Prefix incorrect"
 
-let identifier :=
-    | ~ = identifier_core; ~ = qualifiers; EOF; { identifier_core, qualifiers }
-                                              | ~ = identifier_core; _blah = identifier_core; ~ = qualifiers; EOF; { identifier_core, qualifiers }
+    else (
+      let obj_id = String.lowercase_ascii hash in
+
+      String.iter (fun c ->
+        match c with
+        | 'a'..'f' | '0'..'9' -> ()
+        | _ -> failwith "Object_ID is invalid"
+      ) hash;
+
+      if String.length obj_id = 40 then
+        let obj_id = Array.init 40 (String.get obj_id) in
+        match obj_t with
+        | "snp" ->
+          (scheme_v, Snapshot, obj_id)
+        | "rel" ->
+          (scheme_v, Release, obj_id)
+        | "rev" ->
+          (scheme_v, Revision, obj_id)
+        | "dir" ->
+          (scheme_v, Directory, obj_id)
+        | "cnt" ->
+          (scheme_v, Content, obj_id)
+        | _ ->
+          failwith "Object_type invalid"
+      else
+        failwith "Object_ID has wrong length")
+  }
+
+let fragment_qualifier :=
+    | lines = STR; EQUAL; l1 = INT; l2 = option(preceded(DASH, INT));
+  {
+    if String.equal lines "lines" then
+      (l1, l2)
+    else
+      failwith "Fragment_qualifier non-compliant"
+  }
+
+(* TODO
+ *
+ * I. Parsing
+ * ORIGIN URL_ESCAPED
+ * Make compliant with (* RFC 3987 IRI *)
+ *
+ * PATH ABSOLUTE_ESCAPED
+ * Make compliant with (* RFC 3987 absolute path *)
+ *
+ * II. Error handling
+ * Make custom catchable errout
+ *
+ * *)
+let context_qualifier :=
+    | path = STR; EQUAL; pathabs = STR ;
```

Comment:
rename path, something like `ctx_type` would be better (same for other cases: path, origin, visit_or_anchor